Object ttl command that does not change the object idletime

trauzti opened the pull request:
The regular TTL command sets the idletime to zero. This command returns the TTL in seconds without modifying the idletime.
This can be useful when you are fetching information about objects but don't want to touch them at the same time.

steinnes commented:
Nice, this will help with some of the maintenance work we've been doing at the office, when we want to see the TTL of a key without modifying the idletime, to for example figure out the average TTL on inactive objects, without contaminating their "inactivity".

trauzti commented:
Good steinnes! Perhaps more Redis users would find this useful too. I don't see a good reason for the TTL command to change the idletime. Perhaps it would be better to update the TTL command.

It is fairly simple:
By removing the call to lookupKeyRead within ttlGenericCommand, the regular TTL command would leave the idletime unchanged. The reason is that lookupKeyread calls lookupKey which updates the timestamp.

Natim commented:
I am actually looking for something like that. I would need also something like `object serializedlength key`

Natim commented:
@trauzti Apparently you added another commit to your branch that has nothing to do with this PR.

trauzti commented:
Thanks @Natim will put my new commit into another branch :+1: (Done)

Natim commented:
@trauzti Can you add a test as well please?

mattsta commented:
> I would need also something like object serializedlength key

don't forget about DEBUG OBJECT:

``` haskell
127.0.0.1:6379> keys *
1) "bar"
2) "abc"
3) "foo"
4) "counter"
5) "mylist"
6) "TEST"
7) "key"
127.0.0.1:6379> debug object TEST
Value at:0x7f9e0c100040 refcount:1 encoding:ziplist serializedlength:16 lru:11273344 lru_seconds_idle:15
127.0.0.1:6379> 1000 lpush bigkey abcdef
127.0.0.1:6379> debug object bigkey
Value at:0x7f9e09c095e0 refcount:1 encoding:quicklist serializedlength:122 lru:11273401 lru_seconds_idle:6 ql_nodes:1 ql_avg_node:1000.00 ql_ziplist_max:-2 ql_compressed:0 ql_uncompressed_size:8011
```

Note: to get `serializedlength`, Redis actually serializes out your data, but then throws out the result, to figure out how big it will be.

steinnes commented:
Interesting, does `lru_seconds_idle` from `DEBUG` differ from the idletime @trauzti is inspecting?

mattsta commented:
DEBUG OBJECT gives **idle time** in _seconds_ based on the LRU (**not** the expire time!) using:

``` c
estimateObjectIdleTime(val)/1000
```

where `estimateObjectIdleTime()` is:

``` c
/* Given an object returns the min number of milliseconds the object was never
 * requested, using an approximated LRU algorithm. */
unsigned long long estimateObjectIdleTime(robj *o) {
    unsigned long long lruclock = LRU_CLOCK();
    if (lruclock >= o->lru) {
        return (lruclock - o->lru) * REDIS_LRU_CLOCK_RESOLUTION;
    } else {
        return (lruclock + (REDIS_LRU_CLOCK_MAX - o->lru)) *
                    REDIS_LRU_CLOCK_RESOLUTION;
    }
}
```

The code in this PR wants reads remaining **expire** time (without updating it) using:

``` haskell
return [expire timestamp for key] - [current timestamp];
```

trauzti commented:
Just commited a test.

Natim commented:
Thanks

trauzti commented:
@steinnes @mattsta 

DEBUG OBJECT and OBJECT IDLETIME both call `estimateObjectIdleTime(val)/1000` command so the  `lru_seconds_idle` value is exactly the same as the value from OBJECT IDLETIME

DEBUG OBJECT:

``` addReplyStatusFormat(c,
            "Value at:%p refcount:%d "
            "encoding:%s serializedlength:%lld "
            "lru:%d lru_seconds_idle:%llu",
            (void*)val, val->refcount,
            strenc, (long long) rdbSavedObjectLen(val),
            val->lru, estimateObjectIdleTime(val)/1000);
```

OBJECT IDLETIME:

```
else if (!strcasecmp(c->argv[1]->ptr,"idletime") && c->argc == 3) {
        if ((o = objectCommandLookupOrReply(c,c->argv[2],shared.nullbulk))
                == NULL) return;
        addReplyLongLong(c,estimateObjectIdleTime(o)/1000);
}
```

trauzti commented:
@mattsta 
I just added a command: object serializedlength key
Also added a test for the command on strings. Should we add test cases for other data types?

mattsta commented:
If we're adding all fields from `DEBUG OBJECT` to `OBJECT [field] [key]` one-by-one, we could allow `OBJECT DEBUG [key]` to return the current result of `DEBUG OBJECT [key]` (then we could completely remove `DEBUG OBJECT`).

**plus** for breaking even more backwards compatibility, we could replace `TYPE` with `OBJECT TYPE`.  Currently `OBJECT` only gives encoding, but the encoding is unrelated to the type in some situations.

trauzti commented:
Well ... The main purpose of this pull request was to add OBJECT TTL which
is not provided from DEBUG OBJECT.
On Jan 6, 2015 7:01 PM, "Matt Stancliff" notifications@github.com wrote:

> If we're adding all fields from DEBUG OBJECT to OBJECT [field] [key]
> one-by-one, we could allow OBJECT DEBUG [key] to return the current
> result of DEBUG OBJECT [key](then we could completely remove DEBUG OBJECT).
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/antirez/redis/pull/2090#issuecomment-68913990.

steinnes commented:
I agree with @trauzti -- we would like to be able to check the remaining TTL without changing the LRU idle time.  We are using redis with ttl'd keys heavily for caching and being able to non-destructively (ie. not touching the "idle time) check the TTL would be very helpful for our use cases.

@mattsta I see the seralizedlength stuff is redundant, but the TTL inspection is new.  Do you feel that inspection also belongs somewhere else, or would you agree this is a meaningful contribution if we remove "serializedlength" key?

Natim commented:
Le 21 févr. 2015 14:42, "Steinn Eldjárn Sigurðarson" <
notifications@github.com> a écrit :

> I agree with @trauzti -- we would like to be able to check the remaining
> TTL without changing the LRU idle time. We are using redis with ttl'd keys
> heavily for caching and being able to non-destructively (ie. not touching
> the "idle time) check the TTL would be very helpful for our use cases.

:+1: same here we have the exact same need @Mozilla.

antirez commented:
Hello, a few comments about this. Please don't implement what I propose ASAP, the idea is to discuss and then when we reach a design, implement it.
1. I feel like we should go the other route of making `TTL` not able to change the idle time of the object. 
2. The list of commands that don't alter the object idle time should be documented.
3. We should also add a `TOUCH` command that just changes the object idle time, so it is always possible to alter the idle time when needed.
4. `DEBUG object` should never be mixed with other APIs since we are free to break `DEBUG` every time we need. The remaining API instead is fixed and is supposed to be more clean.
5. Other commands that should not touch the idle time of the object should be identified.
6. To revert to low-level APIs to avoid touching the object idle time is not a good solution, we should touch the lookup API in order to make this a more natural process.

Note that because the above stuff are very near to the deep-core of Redis, I would like to do the changes myself in order to save time in review & possible reimplementation of the same stuff. The important thing is to reach a design, not to implement the changes, that are pretty easy. Comments welcomed.

steinnes commented:
Great response @antirez.  If anyone is relying on idle time changing due to `TTL` touches, they would just have to rely on an explicit `TOUCH`.

Any idea when these changes might make it into the source? I assume there is a little more process than just the discussion on this GitHub issue, but either way I'll be happy :-)

trauzti commented:
I agree 100% @antirez. The `TTL` command should not be able to change the idle time of the object.

Adding a `TOUCH` command is a good solution and also a good feature that is already in memcached.

quaelin commented:
Hi @antirez, adding a +1 here to the idea of making `TTL` not touch idle time.  This would be extremely useful for maintenance.

antirez commented:
I'm working at this issue, already implemented the low level API but can't decide if to go for the route I myself specified above, or for a "TTL" option "NOTOUCH" so that we can make this backward compatible and merge into 3.2. Any hint? The original approach I proposed is more future proof, the latter is more pragmatic in the short time.

antirez commented:
There is probably a better option, I don't think there is **a single user** that relies on the fact TTL changes the last used time of a key. We can just break compatibility and move forward.

antirez commented:
p.s. I also changed the behavior of `TYPE`.

antirez commented:
All implemented here -> 41d804d

Please comment if you can in the next hours with your thoughts, I'm in the process of releasing 3.2.1 in the next days (end of the week likely).

steinnes commented:
👍  from my perspective @antirez and I'm sure @trauzti as well :-)

trauzti commented:
Good change! Looks good to me.

On Tuesday, 14 June 2016, Steinn Eldjárn Sigurðarson <
notifications@github.com> wrote:

> 👍 from my perspective @antirez https://github.com/antirez and I'm sure
> @trauzti https://github.com/trauzti as well :-)
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> https://github.com/antirez/redis/pull/2090#issuecomment-225887293, or mute
> the thread
> https://github.com/notifications/unsubscribe/AAwbfQoE-j2bVuArqQ68zyGqzEzV2xiYks5qLrJUgaJpZM4CxoWI
> .

antirez commented:
Thanks! Merged into 3.2 branch (for 3.2.1) as well.
